### ISSUE TITLE
Credentials do not renew automatically

### ISSUE THREAD

**Issue by TheEdgeOfRage** [duckdb/duckdb_aws#26]
I use k8s [ServiceAccounts that assume IAM roles using OIDC](https://docs.aws.amazon.com/eks/latest/userguide/associate-service-account-role.html) and run DuckDB inside a k8s pod with the proper service account set. On startup, I have the following commands sent to DuckDB:

```
INSTALL 'httpfs'
LOAD 'httpfs'
INSTALL 'aws'
LOAD 'aws'
CALL load_aws_credentials()
```

Our pod then serves requests and fetches data from S3 on request, meaning that it might be asked to get data from S3 much later compared to pod startup. The problem is that the STS token has expired by then, so DuckDB fails to get the data from S3. What's the best approach to renew these credentials, or ideally, could this plugin be updated to renew them automatically once they expire?

**Comment by samansmink** [comment]
Thanks for reporting, token refresh is not yet supported and definitely something to consider

**Comment by CrashLaker** [comment]
hi @samansmink ,
could you please in a little bit more of detail how does duckdb's **threads** uses the credentials when running on a single machine?
this question relates to the question i've asked here https://github.com/duckdb/duckdb/discussions/10996

in there i've 1 maquine spawning many threads to read many S3 objects. my problem is that this breaks as `duckdb.duckdb.HTTPException: HTTP Error: HTTP GET error on 'https://bucket/prefix/filename.json' (HTTP 500)` after some time.

my question is. when each thread starts. does it fetch the credentials **only once** to then load the files.
or I could for example run another process in background that updates the credentials in `~/.aws/config`

regards,c.

**Comment by samansmink** [comment]
Hi @CrashLaker sure!

> does it fetch the credentials only once to

yes, it does this once, on secret creation.

I'm not sure what the 500 error is, AWS docs are not very clear: https://repost.aws/knowledge-center/http-5xx-errors-s3 could maybe be a throttling thing?

You could try messing with the `http_retries` settings here https://duckdb.org/docs/configuration/overview#configuration-reference

**Comment by CrashLaker** [comment]
hi @samansmink ,

thank you for your reply.
actually after I sent you the message I realized that the 500 status code wasn't actually due to expiration (that would be 403) and then I was trying to find ways to tweak that exponential backoff setting.

you've pointed me to the right place in docs. thank you so much! i'll try.

regards,c.

**Comment by valerydluski** [comment]
@samansmink Are there any updates on this issue?

**Comment by samansmink** [comment]
No, sorry i can't give a timeline for automatic credential refreshing at this point

**Comment by vfilter** [comment]
We've been experiencing a similar issue where we have a regular ETL batch job on EKS k8s cron that is very long-running (>3h) and at some point duckdb fails to authenticate against S3 despite earlier rows in the batch sql script working flawlessly. 

Is anyone aware of a workaround for this issue?

**Comment by samansmink** [comment]
We are working towards a refresh mechanism, see https://github.com/duckdb/duckdb-httpfs/pull/21

You can try this out if you install httpfs and aws using:

```
force install aws from core_nightly;
force install httpfs from core_nightly;
```

**Comment by tlinhart** [comment]
@samansmink is this supposed to be working already? I'm still having problems with it even if I specify the `REFRESH auto` option when creating S3 secret.

For the context. I have an AWS Batch job running on ECS/Fargate. The underlying ECS task is granted temporary credentials for its task role with the expiration of 6 hours. This can be checked by ECS execing into the task and calling

```shell
curl http://169.254.170.2$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI
```

The job's application is a long-running Python ETL which processes a large amount of data and periodically spills Parquet files on S3. Before using the `REFRESH auto` option, after the 6 hours the application ended with an exception when trying to write a Parquet file on S3:

```
HTTPException: HTTP Error: Unable to connect to URL 
"https://<redacted>.s3.amazonaws.com/<redacted>/data.parquet": 400 (Bad Request).
```

I then tried to provide the refresh option:

```python
duckdb.sql("CREATE SECRET (TYPE s3, PROVIDER credential_chain, REFRESH auto)")
```

However, now after the 6 hours the application just hangs without erroring. I log a message right before writing the Parquet file (via `COPY ... TO '<redacted>/data.parquet' (FORMAT parquet, COMPRESSION zstd)`) and I can see here the application just stops and hangs forever (until the job timeout).

**Comment by tlinhart** [comment]
Is there something I can do to help debug the issue? Turn on verbose logging or try something else.

**Comment by sudonickk** [comment]
@prakashbalaji were you able to fix the issue?

**Comment by ljyanesm** [comment]
@sudonickk, I'm currently on the latest version of `duckdb` (1.4.1) experiencing the same issue using EKS Pod Identity as both @tlinhart, and @prakashbalaji.

@samansmink, not sure how to reproduce it faster than this, but if you launch a pod on eks and leave it running long enough (~6h in my case) it'll start happening.

**Comment by ruiyang2015** [comment]
in duckdb 1.4.1 try to use following secret in ec2 instance with oidc:
```
CREATE OR REPLACE SECRET secret (
    TYPE s3,
    PROVIDER credential_chain,
    CHAIN 'sts',
    ASSUME_ROLE_ARN 'my_role',
    validation 'none',
    REFRESH AUTO
);
```
and still get 400 bad request immediately. it is not even load for the first time.

**Comment by ruiyang2015** [comment]
using 
```
CREATE OR REPLACE SECRET secret (
    TYPE s3,
    PROVIDER credential_chain,
    REFRESH AUTO
);
```
give validation error in duckdb 1.4.1 now

**Comment by ruiyang2015** [comment]
> [@sudonickk](https://github.com/sudonickk), I'm currently on the latest version of `duckdb` (1.4.1) experiencing the same issue using EKS Pod Identity as both [@tlinhart](https://github.com/tlinhart), and [@prakashbalaji](https://github.com/prakashbalaji).
> 
> [@samansmink](https://github.com/samansmink), not sure how to reproduce it faster than this, but if you launch a pod on eks and leave it running long enough (~6h in my case) it'll start happening.

can you share how do you create the s3  secret to make sts work?

**Comment by ljyanesm** [comment]
Sure @ruiyang2015 , I am using

```
    CREATE OR REPLACE SECRET secret (
        TYPE s3,
        PROVIDER credential_chain,
        REFRESH AUTO
    );
```

**Comment by coste-v** [comment]
Hello @ruiyang2015 !
Did you managed to use @ljyanesm snippet and make it work ?
Thanks !

**Comment by ruiyang2015** [comment]
> Hello [@ruiyang2015](https://github.com/ruiyang2015) ! Did you managed to use [@ljyanesm](https://github.com/ljyanesm) snippet and make it work ? Thanks !

still not working for me

**Comment by ljyanesm** [comment]
In case it helps anyone, I've ended up refreshing the creds application side when they expire by checking the exception, recreating the creds and trying again.

We'll use this until the PR above is merged.

**Comment by ronin1** [comment]
We're affected by this bug via [Shaper](https://github.com/taleshape-com/shaper/issues/138) & has to use rolling container restart or repeated secret save (schedule task) intervals to work around this problem.

**Comment by jcolot** [comment]
> In case it helps anyone, I've ended up refreshing the creds application side when they expire by checking the exception, recreating the creds and trying again.
> 
> We'll use this until the PR above is merged.

Thanks for the tip! 

Do you still recreate the secret with `PROVIDER credentials_chain` or do you inject credentials obtained by an AWS SDK like boto3?

Recreating the secret with `PROVIDER credentials_chain` seems to fail after some time, I'm on 1.4.4.

**Comment by ljyanesm** [comment]
@jcolot , just recreate the credentials as above. 

It may be that this is not an issue in the environment we are running the ddb instance that was getting expired credentials and may be an issue in a different environment (i.e locally).
